cpp/include/cudf/lists/detail/scatter.cuh

```diff
@@ -0,0 +1,727 @@
+        bound_list_row.size(),
+        [input_list_start,
+         output_start_offset,
+         label,
+         d_child_list_views,
+         d_child_offsets =
+           child_offsets.template data<int32_t>()] __device__(auto const& child_list_index) {
+          auto child_start_idx = d_child_offsets[input_list_start + child_list_index];
+          auto child_end_idx   = d_child_offsets[input_list_start + child_list_index + 1];
+
+          d_child_list_views[output_start_offset + child_list_index] = unbound_list_view{
+            label, input_list_start + child_list_index, child_end_idx - child_start_idx};
+        });
+    };
+
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+                       thrust::make_counting_iterator<size_type>(0),
+                       list_vector.size(),
+                       populate_child_list_views);
+
+    // child_list_views should now have been populated, with source and target references.
+
+    auto begin = thrust::make_transform_iterator(
+      child_list_views.begin(), [] __device__(auto const& row) { return row.size(); });
+
+    auto child_offsets = cudf::strings::detail::make_offsets_child_column(
+      begin, begin + child_list_views.size(), mr, stream);
+
+    auto child_column =
+      cudf::type_dispatcher(source_lists_column_view.child().child(1).type(),
+                            list_child_constructor{},
+                            child_list_views,
+                            child_offsets->view(),
+                            cudf::lists_column_view(source_lists_column_view.child()),
+                            cudf::lists_column_view(target_lists_column_view.child()),
+                            mr,
+                            stream);
+
+    auto child_null_mask =
+      source_lists_column_view.child().nullable() || target_lists_column_view.child().nullable()
+        ? construct_child_nullmask(
+            list_vector, list_offsets, source_lists, target_lists, num_child_rows, mr, stream)
+        : std::make_pair(rmm::device_buffer{}, 0);
+
+    return cudf::make_lists_column(num_child_rows,
+                                   std::move(child_offsets),
+                                   std::move(child_column),
+                                   child_null_mask.second,            // Null count
+                                   std::move(child_null_mask.first),  // Null mask
+                                   stream,
+                                   mr);
+  }
+};
+
+/**
+ * @brief Checks that the specified columns have matching schemas, all the way down.
+ */
+void assert_same_data_type(column_view const& lhs, column_view const& rhs)
+{
+  CUDF_EXPECTS(lhs.type().id() == rhs.type().id(), "Mismatched Data types.");
+  CUDF_EXPECTS(lhs.num_children() == rhs.num_children(), "Mismatched number of child columns.");
+
+  for (int i{0}; i < lhs.num_children(); ++i) { assert_same_data_type(lhs.child(i), rhs.child(i)); }
```

Comment:
Hmmm, `column_view` needs a `child_begin()` and `child_end()` so this can be a `std::all_of` call.

Reply:
A couple of problems there:
1. I can't seem to find a version of `std::all_of()` that takes two sequences. Pardon me if I've missed it.
2. The predicate in question (i.e. `assert_same_data_type()`) isn't actually a boolean predicate. It errors out via `CUDF_EXPECTS`.

I can convert this into a `std::for_each()` on a counting iterator, but that might just be window dressing.

Reply:
Good points.